File: 0189-rotate-array/0189-rotate-array.py
class Solution(object):
    def rotate(self, nums, k):
        '''for i in range(k):
            temp=nums[len(nums)-1]
            for j in range((len(nums)-1),0,-1):
                nums[j]=nums[j-1]
            nums[0]=temp
        return nums'''

        if len(nums)<1:
            return nums


        n = len(nums)
        k%=n

        
        def reverse(l, r):
            while l < r:
                nums[l], nums[r] = nums[r], nums[l]
                l += 1
                r -= 1

        reverse(0, n - 1)
        reverse(0, k - 1)
        reverse(k, n - 1)

        return nums

File: 0189-rotate-array/0189-rotate-array.java
class Solution {

    public void rev(int first, int last,int[] arr) {
        
        

        while(first<last){
            int temp = arr[first];
            arr[first]=arr[last];
            arr[last]= temp;

            first++;
            last--;
        }
        
    }
    
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k%nums.length;

        rev(0,n-k-1,nums);
        rev(n-k,nums.length-1,nums);
        rev(0,nums.length-1,nums);
    }
}
